### ISSUE TITLE
`okta_policy_rule_mfa` : Creating multiple resources with priority causing errors

### ISSUE THREAD

**Issue by BalaGanaparthi** [okta/terraform-provider-okta#1059]
### Community Note

- Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
- Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
- If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

```version
Terraform v1.1.4
on linux_amd64
+ provider registry.terraform.io/okta/okta v3.23.0
```

### Error Description
```error
When multiple resources with priority are created, it seems the resources are created out of order (like the higher priority resources getting created prior to lower priority) and resulting in panic.
 
Resources are created in a loop 

A related issue is, when creating resources in loop and priority are not assigned the created resources at times gets assigned the same priority #.. like the two resources getting the prioriy of 1
```

### Affected Resource(s)

- okta_policy_rule_mfa

### Terraform Configuration Files

`variables.tf`
```hcl
mfa_enroll_rules = [
                {
                    name = "MFA Policy App-A Rule-1",
                    priority =1,
                    status="ACTIVE",
                    enroll="LOGIN",
                    network_connection="ANYWHERE",
                    network_includes = [],
                    network_excludes = []
                },
                {
                    name = "MFA Policy App-A Rule-2",
                    priority = 2,
                    status="ACTIVE",
                    enroll="LOGIN",
                    network_connection="ANYWHERE",
                    network_includes = [],
                    network_excludes = []
                },
                {
                    name = "MFA Policy App-A Rule-3",
                    priority = 3,
                    status="ACTIVE",
                    enroll="LOGIN",
                    network_connection="ANYWHERE",
                    network_includes = [],
                    network_excludes = []
                }
]
```
`main.tf`
```hcl
resource "okta_policy_rule_mfa" "mfa_enroll_policy_rule" {
  for_each = { for mepr in local.mfa_rules : mepr.policy-rule-key => mepr }

  policy_id = each.value.policy_id
  name      = each.value.rule.name
  status    = each.value.rule.status
  enroll    = each.value.rule.enroll
  priority  = each.value.rule.priority
}
```

### Plan Output

```plan
# module.mfa_enrollment_policies_rules.okta_policy_rule_mfa.mfa_enroll_policy_rule["00p3gi95rd5vWoXLe1d7.MFA Policy App-A Rule-2"] will be created
  + resource "okta_policy_rule_mfa" "mfa_enroll_policy_rule" {
      + enroll             = "LOGIN"
      + id                 = (known after apply)
      + name               = "MFA Policy App-A Rule-2"
      + network_connection = "ANYWHERE"
      + policy_id          = "00p3gi95rd5vWoXLe1d7"
      + priority           = 2
      + status             = "ACTIVE"
    }

  # module.mfa_enrollment_policies_rules.okta_policy_rule_mfa.mfa_enroll_policy_rule["00p3gi95rd5vWoXLe1d7.MFA Policy App-A Rule-3"] will be created
  + resource "okta_policy_rule_mfa" "mfa_enroll_policy_rule" {
      + enroll             = "LOGIN"
      + id                 = (known after apply)
      + name               = "MFA Policy App-A Rule-3"
      + network_connection = "ANYWHERE"
      + policy_id          = "00p3gi95rd5vWoXLe1d7"
      + priority           = 3
      + status             = "ACTIVE"
    }

  # module.mfa_enrollment_policies_rules.okta_policy_rule_mfa.mfa_enroll_policy_rule["00p3gi9oylAqgSglg1d7.MFA Policy App-A Rule"] will be created
  + resource "okta_policy_rule_mfa" "mfa_enroll_policy_rule" {
      + enroll             = "LOGIN"
      + id                 = (known after apply)
      + name               = "MFA Policy App-A Rule"
      + network_connection = "ANYWHERE"
      + policy_id          = "00p3gi9oylAqgSglg1d7"
      + priority           = 1
      + status             = "ACTIVE"
    }
```

### Panic Output

```panic
╷
│ Error: failed to create MFA policy rule: failed to create policy rule: the API returned an error: Internal Server Error, x-okta-request-id=YlUZ_Cu8kkrBfZ0oNfMJ_gAADsU
│ 
│   with module.mfa_enrollment_policies_rules.okta_policy_rule_mfa.mfa_enroll_policy_rule["00p3gi95rd5vWoXLe1d7.MFA Policy App-A Rule-2"],
│   on modules/mfa_enrollment_policies_rules/main.tf line 3, in resource "okta_policy_rule_mfa" "mfa_enroll_policy_rule":
│    3: resource "okta_policy_rule_mfa" "mfa_enroll_policy_rule" {
│ 
╵
╷
│ Error: failed to create MFA policy rule: provided priority was not valid, got: 3, API responded with: 1. See schema for attribute details
│ 
│   with module.mfa_enrollment_policies_rules.okta_policy_rule_mfa.mfa_enroll_policy_rule["00p3gi95rd5vWoXLe1d7.MFA Policy App-A Rule-3"],
│   on modules/mfa_enrollment_policies_rules/main.tf line 3, in resource "okta_policy_rule_mfa" "mfa_enroll_policy_rule":
│    3: resource "okta_policy_rule_mfa" "mfa_enroll_policy_rule" {
│ 
╵
ERROR: 1
```

### Expected Behavior

```
The rules are created within the enrollment policy, with the provided priority
```

### Actual Behavior

```
Getting panic/error during apply
In the okta tenant, 2 of 3 rules are created ("MFA Policy App-A Rule-1" and "MFA Policy App-A Rule-3" are created)
```

### Steps to Reproduce

1. `terraform apply`

### Important Factoids

- Rule with `name="MFA Policy App-A Rule"` got created with `priority=1` : This is as expected 👍  
- Rule with `name="MFA Policy App-A Rule-3"` got created with `priority=2` , The priority was auto corrected at Okta from the original `priority=3`
- Rule with `name="MFA Policy App-A Rule-2"` did not get created 

### References

- #0000

**Comment by bogdanprodanj** [comment]
@BalaGanaparthi, that's a known issue, and I think there was even a ticket created for that error.

**Comment by BalaGanaparthi** [comment]
@bogdanprodanj , There are two issues.
1. When priority is not attached, resources results with overlapping priorities. For this issue there was an issue/jira created but the issue did not resolved (jira was closed) so I had to decompose the resource creation by creating the resources individually insread of using for_each. We may have to reopen the case.
2.  When the priorities are attached, the resources when getting created are getting in out-of `priority` order and getting into issue scenario. From what I observe, my guess of what happened is as follows

- The for_each loop decomposed the resources to be created and associated the priorities as expected to the resources to be created (As shown in the `Plan Output` above) 👍 
- When applying the first resource that got executed at Okta was the one with `priority = 1` and that got created successfully 👍 
- The second resource that got created was with `priority = 3`, that got into an issue at Okta (as one  priority ie, `priority=2` was skipped) and was self corrected by assigning the `priority =2` for the resource and created the resource at Okta 🟧
- The third resource that got created was with `priority = 2`, and this time around there was a resource already created with the auto-corrected  priority which is 2 and that caused a conflict.. and hence the cause of error 👎 

Looks like when priorities are not continuous (skipped a number), then Okta auto corrects and assigns the next highest priority, but causes error when a overlapping priority is encountered (due to the earlier auto correction)

**Comment by monde** [comment]
Great to see @bogdanprodanj ! @BalaGanaparthi thanks for the details. We have an upgrade to the okta-sdk-golang that has enhancements to policies and it is my hope that when the upgrade is complete (in the next week) I'll be able to knock down all of the policy issues that are open.

**Comment by kasirajans** [comment]
@BalaGanaparthi  thanks for the explanation. @monde  this happens for alll the resource which have option with priority. Also I’m planning to order the file with priority = [resource 1, resource 2, resource 3] instead hard coding the value so i wont do any typo as most of the tf code we do copy paste and chances of duplicating the priority are high. Currently we do see lots of duplication happens with priority as we are copy pasting the existing resource and forget to update the values. So doing priority based on the ordered list will help us to go more agile.

**Comment by monde** [comment]
@BalaGanaparthi 0 value integers was fixed in [v3.44.0](https://github.com/okta/terraform-provider-okta/releases/tag/v3.44.0) release. Is this issue still not resolved?

**Comment by tduong10101** [comment]
Hi @monde, 

I'm still getting this issue but for mfa and signon policies. 
Similar to the original example. If a new resource is added and all other priorities are updated on tf. 
We'll get the below error, I think it's because the priorities are getting updated before the new resource is created?

"15-Feb-2024 20:15:49 | Error: failed to update sign-on policy: provided priority was not valid, got: 14, API responded with: 12. See schema for attribute details"

We have to hard coded the priorities because we have mfa and signon polices from multi modules.
